### PR TITLE
Thomasht86/sync retry strategy 429

### PR DESCRIPTION
**Pull request by thomasht86**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

Closing #791 

After @jonmv fixed the async retry-strategy, we decided fixing the sync retry strategy for 429 was highest priority to gain confidence that the upcoming change in feed queue handling will not affect users negatively. 

After a nice tip from @jonmv on how to easily simulate slow feeding, it was also easy to add integration test for this. 
These tests would have failed with the old strategy, where 429s were only retried 3 times, and are very relevant as people add slow embedders etc. 

- Also a small refactor to get rid of `WORK_DIR` env variable.

Will create a new issue to add async feeding to the same integration-tests. 

Really don't like to break the DRY-principle by duplicating a lot of code for both docker and cloud tests. 
It would be very nice to refactor the Test classes properly, to make it easier to add both Cloud and Docker applications without duplicating code.

### CODE REVIEW

File: vespa/application.py [hunk 1]
```diff
@@ -802,6 +796,35 @@ def get_document_v1_path(
...
+        )
+        self.retry_strategy = retry_strategy
+
+    def send(self, request, **kwargs) -> Response:
+        for attempt in range(self.num_retries_429):
+            response = super().send(request, **kwargs)
+            if response.status_code == 429:
+                wait_time = (
+                    2** attempt + random.uniform(0, 1)
+                )  # Exponential backoff with jitter. The 10th retry will sleep for 1024 seconds.
```
Comment (jonmv) [inline vespa/application.py:821]: I think that is a bit excessive, perhaps cap it at 10s or so, and even start a bit lower?

**Review by jonmv — DISMISSED** [review]
Nice! I think we should probably avoid sleeping for more than 10s, as the user may just think the feeder is hanging.

File: vespa/application.py (hunk 1 above)
Comment (jonmv) [reply vespa/application.py:823]: Did we decide to reduce this a bit, though? I can't remember now 😁

File: vespa/application.py [hunk 2]
```diff
@@ -802,8 +798,39 @@ def get_document_v1_path(
...
+            allowed_methods=["POST", "GET", "DELETE", "PUT"],
+        )
+        self.retry_strategy = retry_strategy
+
+    def send(self, request, **kwargs) -> Response:
+        for attempt in range(self.num_retries_429):
+            response = super().send(request, **kwargs)
+            if response.status_code == 429:
+                wait_time = (
+                    2** attempt + random.uniform(0, 1)
```
Comment (jonmv) [inline vespa/application.py:822]:
```suggestion
                    0.1 * (1.618** attempt) + random.uniform(0, 1)
```

**Review by jonmv — APPROVED** [review]
Great!

File: vespa/application.py (hunk 2 above)
Comment (thomasht86) [reply vespa/application.py]: Had a commit, but got distracted from pushing it. 
Great that you saw it! 👍

**Review by jonmv — APPROVED** [review]

**Pull request merged**